### PR TITLE
fix: enable native_datafusion Spark SQL tests for #3320, #3401, #3719

### LINKED ISSUE #3311: [native_datafusion] [Spark SQL Tests] Schema incompatibility tests expect exceptions that native_datafusion handles gracefully
## Summary

Several Spark SQL tests fail because they expect a `SparkException` when reading Parquet with mismatched schemas, but `native_datafusion` either handles type widening/mismatches gracefully or throws errors with different exception types/messages than Spark expects.

### Original 8 tests (from initial report)

| Test | Suite | Root Cause | Difficulty |
|------|-------|------------|------------|
| SPARK-35640: read binary as timestamp should throw schema incompatible error | `ParquetIOSuite` | Native reader doesn't reject binary→timestamp | **Easy** — add type check in schema adapter |
| SPARK-35640: int as long should throw schema incompatible error | `ParquetIOSuite` | Native reader allows INT32→INT64 widening | **Easy** — add type check in schema adapter |
| SPARK-36182: can't read TimestampLTZ as TimestampNTZ | `ParquetQuerySuite` | INT96 timestamps don't carry timezone info in Parquet schema, so native reader can't detect LTZ→NTZ mismatch | **Hard** — would need INT96-specific handling |
| SPARK-34212: Parquet should read decimals correctly | `ParquetQuerySuite` | Different decimal precision/scale handling | **Medium** — needs decimal-specific validation |
| SPARK-45604 / schema mismatch failure error message | `ParquetSchemaSuite` | Tests check for `SchemaColumnConvertNotSupportedException` which is a Spark-internal type | **Ignore** — can't match Spark-internal exception types from native |
| Spark native readers should respect spark.sql.caseSensitive | `FileBasedDataSourceSuite` | Tests check for `SparkRuntimeException` with specific error class | **Ignore** — can't match Spark error classes from native |
| row group skipping doesn't overflow when reading into larger type | `ParquetReadSuite` | Already fixed on main | N/A |

### Additional tests discovered during PR #3416

| Test | Suite | Root Cause |
|------|-------|------------|
| SPARK-25207: exception when duplicate fields in case-insensitive mode | `ParquetFilterSuite` (x2: V1 and V2) | Exception wrapping differs — Comet wraps in `SparkException` but the cause is not `RuntimeException` as test expects |
| SPARK-26709: OptimizeMetadataOnlyQuery does not handle empty records correctly | `SQLQuerySuite` | Schema validation rejects INT32→INT64 coercion (empty file written with `lit(1)` as INT32, but `range()` writes INT64; combined read fails) |

## Analysis from PR #3416

PR #3416 attempted to add Spark-compatible schema validation in the native schema adapter. Key learnings:

### Approach taken
- Added validation in `SparkSchemaAdapter.map_schema()` (Rust) that rejects type coercions Spark's vectorized reader would reject
- New config `spark.comet.parquet.schemaValidation.enabled` (default: true)
- Wrapped native schema errors in `SparkException` with compatible error messages in `CometExecIterator`
- Passed `schema_evolution_enabled` to native side via proto

### What worked
- 2 tests were unignored (SPARK-35640 binary→timestamp and int→long)
- Schema validation correctly catches most type mismatches

### What didn't work (and why the PR was closed)
- **More tests were ignored (3) than unignored (2)**, making it a net negative
- **INT96 timestamp limitation**: Arrow-rs represents INT96 timestamps as `Timestamp(Nanosecond, None)` — same representation as TimestampNTZ. The native reader cannot distinguish "this was originally LTZ written as INT96" from "this is NTZ", making SPARK-36182 unfixable without INT96-specific metadata tracking
- **Exception type mismatch**: Spark tests check for specific exception cause types (`RuntimeException`, `SchemaColumnConvertNotSupportedException`, `SparkRuntimeException` with error classes) that can't be produced from native Rust code
- **Schema evolution interaction**: Some tests write data with mixed types across partitions (INT32 from `lit(1)` + INT64 from `range()`), and the validation correctly rejects INT32→INT64 on Spark 3.x where schema evolution is off — but this breaks tests that relied on silent widening

### Bug found and fixed
The `fileNotFoundPattern` regex in `CometExecIterator` had a bug: it included the `^External:` prefix in the pattern, but the code stripped that prefix before matching, so it could never match. This caused 4 `HiveMetadataCacheSuite` tests to fail because file-not-found errors weren't being wrapped in `SparkException` with `FileNotFoundException`. Fix: [320cc02](https://github.com/apache/datafusion-comet/commit/320cc02b3).

## Recommended approach

Rather than adding schema validation on the native side, it may be simpler to:

1. **Ignore tests that check Spark-internal exception types** — these can never be matched from native code
2. **Add targeted type checks** for the 2 easy cases (binary→timestamp, int→long) without a full validation framework
3. **Fix the `CometExecIterator` regex bug** independently (the fix from PR #3416 should be cherry-picked)
4. **Accept** that INT96 timestamp LTZ/NTZ detection is a known limitation of the native reader

## Related

- PR #3416 (closed): Full schema validation implementation
- Discovered in CI for #3307 (enable native_datafusion in auto scan mode)

### LINKED ISSUE #3320: [native_datafusion] [Spark SQL Tests] CometNativeExec crash — executed without serialized plan
## Summary

Several tests crash with `CometRuntimeException: CometNativeExec should not be executed directly without a serialized plan` when native_datafusion is used in auto mode.

## Failing Tests

- `ParquetFilterSuite`: "SPARK-31026: Parquet predicate pushdown for fields having dots in the names"
- `ParquetFilterSuite`: "Filters should be pushed down for Parquet readers at row group level"
- `ParquetFilterSuite`: "filter pushdown - StringPredicate"
- `FileBasedDataSourceSuite`: "SPARK-41017: filter pushdown with nondeterministic predicates"

## Error

```
CometRuntimeException: CometNativeExec should not be executed directly without a serialized plan:
CometNativeScan parquet [...] PushedFilters: [IsNotNull(...), EqualTo(...)], ...
```

## Root Cause

These tests execute queries that go through a code path where `CometNativeScan` is included in a plan but the plan serialization to the native side didn't happen (likely because the plan was executed in a test-specific way that bypasses the normal `CometSparkSessionExtensions` hooks). The `CometNativeScan` node's `doExecuteColumnar` throws because it requires a serialized native plan.

This may also indicate cases where `CometNativeScan` is placed in a plan that isn't fully wrapped by `CometExec`, possibly because filter pushdown inspection tests run plans in a non-standard way.

## Related

Discovered in CI for #3307 (enable native_datafusion in auto scan mode).

### LINKED ISSUE #3401: [native_datafusion] [Spark SQL Tests] Streaming self-union returns no results
## Description

When running Spark SQL tests with `native_datafusion` scan enabled, two streaming self-union tests in `StreamingSelfUnionSuite` fail because the streaming query result is `None` instead of containing the expected data.

## Failing Tests

- `self-union, DSv1, read via DataStreamReader API`
- `self-union, DSv1, read via table API`

## Error

```
Assert on query failed: : None was empty
```

## Context

Found during CI run: https://github.com/apache/datafusion-comet/actions/runs/21691646806/job/62552819424?pr=3393

### LINKED ISSUE #3719: native_datafusion: CometNativeException not converted to SparkRuntimeException for case-sensitivity error
## Description

When running Spark SQL tests with `spark.comet.scan.impl=native_datafusion`, the `FileBasedDataSourceSuite` test "Spark native readers should respect spark.sql.caseSensitive" fails with:

```
java.lang.ClassCastException: class org.apache.comet.CometNativeException cannot be cast to class org.apache.spark.SparkRuntimeException
```

The test expects a `SparkRuntimeException` when case-sensitive mode is enabled and the query references a column with different casing. DataFusion produces a `CometNativeException` instead.

## Affected Test

- `FileBasedDataSourceSuite`: "Spark native readers should respect spark.sql.caseSensitive - parquet"

## Expected Behavior

The native error should be converted to a `SparkRuntimeException` so Spark error handling works correctly.

## Parent Issue

Split from #3311.

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes #3320.
Closes #3401.
Closes #3311.
Closes #3719.

## Rationale for this change

Several Spark SQL tests were skipped with `IgnoreCometNativeDataFusion` but now pass, either without code changes or with a targeted error conversion fix.

### Enabled 5 tests (tags removed)

**#3320 — ParquetFilterSuite** (2 tests):
- `SPARK-31026: Parquet predicate pushdown for fields having dots in the names`
- `Filters should be pushed down for Parquet readers at row group level`

Both already had conditional assertion guards for native execution. The skip tag was overly conservative.

**#3401 — StreamingSelfUnionSuite** (2 tests):
- `self-union, DSv1, read via DataStreamReader API`
- `self-union, DSv1, read via table API`

Streaming DSv1 Parquet reads work correctly with `native_datafusion`.

**#3719 — FileBasedDataSourceSuite** (1 test):
- `Spark native readers should respect spark.sql.caseSensitive`

Fixed by adding error conversion: DataFusion's "Unable to get field named" schema error is now detected in `throw_exception` and converted to a `SparkRuntimeException` with error class `_LEGACY_ERROR_TEMP_2093` via the JSON error path. The Spark test is updated to assert the error class without checking exact message parameters, since DataFusion reports the field names differently than Spark.

### Error conversion changes

- **`native/spark-expr/src/error.rs`**: Added `SparkError::DuplicateFieldCaseInsensitive` variant with `required_field_name` and `matched_fields` params, error class `_LEGACY_ERROR_TEMP_2093`
- **`native/core/src/errors.rs`**: Added `try_convert_duplicate_field_error()` to detect DataFusion's schema error pattern and convert to the new `SparkError` variant
- **`ShimSparkErrorConverter.scala`** (3.4, 3.5, 4.0): Added handler calling `QueryExecutionErrors.foundDuplicateFieldInCaseInsensitiveModeError()`

### Re-linked 8 tests from #3311 to specific issues

- **#3719** (1 test): CometNativeException not converted to SparkRuntimeException — now fixed
- **#3720** (7 tests): No error thrown for schema mismatch with incompatible types — still open

### Documentation

Updated `docs/source/contributor-guide/parquet_scans.md` to reflect that duplicate field errors in physical Parquet files now produce the correct Spark exception type.

## How are these changes tested?

All 5 enabled tests were run locally against Spark 3.5.8 with `ENABLE_COMET=true ENABLE_COMET_ONHEAP=true -Dspark.comet.scan.impl=native_datafusion` and pass. The diff was regenerated following the documented process and verified to apply cleanly.

### CODE REVIEW

File: native/core/src/errors.rs [hunk 1]
```diff
@@ -474,6 +476,54 @@ fn throw_spark_error_as_json(
     )
 }
 
+/// Try to convert a DataFusion "Unable to get field named" error into a SparkError.
+/// DataFusion produces this error when reading Parquet files with duplicate field names
+/// in case-insensitive mode. For example, if a Parquet file has columns "B" and "b",
+/// DataFusion may deduplicate them and report: Unable to get field named "b". Valid
+/// fields: ["A", "B"]. When the requested field has a case-insensitive match among the
+/// valid fields, we convert this to Spark's _LEGACY_ERROR_TEMP_2093 error.
+fn try_convert_duplicate_field_error(error_msg: &str) -> Option<SparkError> {
```
Comment (andygrove) [inline native/core/src/errors.rs:485]: I'm not sure if this is overkill or not. We could just update the Spark SQL test to allow for different exceptions with Comet

**Comment by andygrove** [comment]
Thanks for the review @mbutrovich!

File: native/core/src/errors.rs (hunk 1 above)
Comment (parthchandra) [reply native/core/src/errors.rs:485]: late comment: 
You're right, this is overkill. We can, if the need arises (as in this case), not convert the `_LEGACY_ ` errors? Or even more broadly, not do this for the errors that originate in `QueryExecutionErrors`. 
The error framework is important for the errors in `org.apache.spark.sql.errors.ExecutionErrors` because those are SQL errors and correspond to some pre-defined error codes in ANSI.  But for `QueryExecutionErrors` we do not have to be strict.

Comment (parthchandra) [reply native/core/src/errors.rs:485]: nvm. `QueryExecutionErrors are also SQL errors.
